pkg/pillar/types/global.go

```diff
@@ -103,312 +109,439 @@ func (gs *GlobalStatus) UpdateItemValuesFromGlobalConfig(gc GlobalConfig) {
-// Zero integers means to use default
-func ApplyGlobalConfig(newgc GlobalConfig) GlobalConfig {
+func (configSpec ConfigItemSpec) ParseSpecToValue() ConfigItemValue {
+	var item ConfigItemValue
+	item.Key = configSpec.Key
+	switch configSpec.ItemType {
+	case ConfigItemTypeBool:
+		item.boolValue = configSpec.BoolDefault
+	case ConfigItemTypeInt:
+		item.intValue = configSpec.IntDefault
+	case ConfigItemTypeString:
+		item.stringValue = configSpec.StringDefault
+	case ConfigItemTypeTristate:
+		item.triStateValue = configSpec.TriStateDefault
+	}
+	return item
+}
 
-	if newgc.ConfigInterval == 0 {
-		newgc.ConfigInterval = GlobalConfigDefaults.ConfigInterval
-	}
-	if newgc.MetricInterval == 0 {
-		newgc.MetricInterval = GlobalConfigDefaults.MetricInterval
-	}
-	if newgc.ResetIfCloudGoneTime == 0 {
-		newgc.ResetIfCloudGoneTime = GlobalConfigDefaults.ResetIfCloudGoneTime
-	}
-	if newgc.FallbackIfCloudGoneTime == 0 {
-		newgc.FallbackIfCloudGoneTime = GlobalConfigDefaults.FallbackIfCloudGoneTime
-	}
-	if newgc.MintimeUpdateSuccess == 0 {
-		newgc.MintimeUpdateSuccess = GlobalConfigDefaults.MintimeUpdateSuccess
-	}
-	if newgc.NetworkGeoRedoTime == 0 {
-		newgc.NetworkGeoRedoTime = GlobalConfigDefaults.NetworkGeoRedoTime
+type ConfigItemSpecMap struct {
+	// Map Key: settingKey
+	// ConfigItemSpec: settingKey
+	GlobalSettings map[GlobalSettingKey]ConfigItemSpec
+	// Map Key: per-agent setting key
+	// ConfigItemSpec.Key: per-agent setting key
+	AgentSettings map[AgentSettingKey]ConfigItemSpec
+}
+
+func (specMap *ConfigItemSpecMap) AddIntItem(key GlobalSettingKey, agentSettingKey AgentSettingKey, defaultInt uint32, min uint32, max uint32) {
+	if defaultInt < min || defaultInt > max {
+		log.Fatalf("Adding int item %s failed. Value does not meet given min/max criteria", key)
 	}
-	if newgc.NetworkGeoRetryTime == 0 {
-		newgc.NetworkGeoRetryTime = GlobalConfigDefaults.NetworkGeoRetryTime
+	configItem := ConfigItemSpec{
+		ItemType:   ConfigItemTypeInt,
+		Key:        string(key),
+		IntDefault: defaultInt,
+		IntMin:     min,
+		IntMax:     max,
 	}
-	if newgc.NetworkTestDuration == 0 {
-		newgc.NetworkTestDuration = GlobalConfigDefaults.NetworkTestDuration
+	if agentSettingKey != "" {
+		configItem.Key = string(agentSettingKey)
+		specMap.AgentSettings[agentSettingKey] = configItem
 	}
-	if newgc.NetworkTestInterval == 0 {
-		newgc.NetworkTestInterval = GlobalConfigDefaults.NetworkTestInterval
+	if key != "" {
+		specMap.GlobalSettings[key] = configItem
 	}
-	// We allow newgc.NetworkTestBetterInterval to be zero meaning disabled
+	log.Debugf("Added int item %s", key)
+}
 
-	if newgc.NetworkFallbackAnyEth == TS_NONE {
-		newgc.NetworkFallbackAnyEth = GlobalConfigDefaults.NetworkFallbackAnyEth
-	}
-	if newgc.NetworkTestTimeout == 0 {
-		newgc.NetworkTestTimeout = GlobalConfigDefaults.NetworkTestTimeout
+func (specMap *ConfigItemSpecMap) AddBoolItem(key GlobalSettingKey, agentSettingKey AgentSettingKey, defaultBool bool) {
```

Comment:
Hmm.. Now that we have separate types for GlobalSettingKey and agentSettingKey, may be its just better to have separate methods to add to GlobalAgentSetting and AgentSettings??
 
func (specMap *ConfigItemSpecMap) AddBoolItem(key GlobalSettingKey, defaultBool bool)

func (specMap *ConfigItemSpecMap) AddAgentSettingBoolItem(key AgentSettingKey, defaultBool bool)

Combining both in the same routine seems confusing.. We expect most items to be just GlobalSetting items..

So essentially, 4 Global setting routines and 1 agent setting routine:

func (specMap *ConfigItemSpecMap) AddIntItem(key GlobalSettingKey, default, min, max int)
func (specMap *ConfigItemSpecMap) AddBoolItem(key GlobalSettingKey, default bool)
func (specMap *ConfigItemSpecMap) AddTriStateItem(key GlobalSettingKey, default TriState)
func (specMap *ConfigItemSpecMap) AddStringItem(key GlobalSettingKey, default string)

And One agent setting Routine ( Since we only have string type agent settings at this point ):

func (specMap *ConfigItemSpecMap) AddAgentSettingStringItem(key AgentSettingKey, default string)